Add Nondeterministic Tol to gradient test in test_modules

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69317
* __->__ #69402

Differential Revision: [D33285781](https://our.internmc.facebook.com/intern/diff/D33285781)

mikaylagawarecki committed:
Add Nondeterministic Tol to gradient test in test_modules

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"




[ghstack-poisoned]

jbschlosser commented on test/test_modules.py:
```diff
@@ -396,6 +397,11 @@ def _test_gradients_helper(self, device, dtype, module_info, check):
         module_cls = module_info.module_cls
         module_inputs = module_info.module_inputs_func(module_info, device=device, dtype=dtype,
                                                        requires_grad=True)
+        # Set some backend-specific validation settings.
+        gradcheck_nondet_tol = 0.0
+        if torch.backends.cudnn.is_available():
+            # cuDNN introduces non-determinism
+            gradcheck_nondet_tol = GRADCHECK_NONDET_TOL
```
After thinking about this some more, I realized it's probably better to set `gradcheck_nondet_tol` on a per-module basis instead of always setting it when cudnn is enabled.

The reason for this is because setting it like this is possibly too liberal, and wouldn't catch certain cases that we do want to catch involving non-determinism. I think the ideal state would be to set the tolerance on a per-module basis only for CUDA tests if cuDNN is enabled, if there's some way to accomplish that. Then we are as strict as possible with the tolerance and only increase it when absolutely necessary. Wdyt about this?

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"




[ghstack-poisoned]

jbschlosser reviewed (approved):
LGTM!

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33285781).

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"


Differential Revision: [D33285781](https://our.internmc.facebook.com/intern/diff/D33285781)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33285781).

mikaylagawarecki committed:
Update on "Add Nondeterministic Tol to gradient test in test_modules"


Differential Revision: [D33285781](https://our.internmc.facebook.com/intern/diff/D33285781)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33285781).
